utils/node-loader/src/task.rs

```diff
@@ -1,41 +1,53 @@
-use std::marker::PhantomData;
+use std::{
+    pin::Pin,
+    sync::{Arc, Mutex},
+};
 
-use anyhow::Result;
+use futures::stream::FuturesUnordered;
 use gear_program::api::Api;
-use tokio::task::JoinHandle;
+use rand::{rngs::SmallRng, seq::SliceRandom, RngCore};
+use tokio::{
+    sync::mpsc::{self, Receiver},
+    task::{JoinError, JoinHandle},
+};
 
 use crate::{args::SeedVariant, reporter::SomeReporter};
 use upload_program::UploadProgramTaskGen;
 
+use self::{
+    generators::{FutureSomeReporter, TaskGen},
+    upload_code::UploadCodeTaskGen,
+};
+
 pub(crate) mod generators;
+mod upload_code;
 mod upload_program;
 
-pub(crate) struct TaskPool<Rng: crate::Rng> {
-    up_task_gen: UploadProgramTaskGen,
-    tasks: Vec<JoinHandle<Result<SomeReporter>>>,
-    _phantom: PhantomData<Rng>,
+type TaskGenVec = Vec<Box<dyn TaskGen<Output = FutureSomeReporter> + Send + Sync>>;
+
+pub(crate) struct TaskPool {
+    rx: Option<Receiver<JoinHandle<SomeReporter>>>,
+    tasks: FuturesUnordered<JoinHandle<SomeReporter>>,
+    seed_gen: Arc<Mutex<Box<dyn RngCore + Send + Sync>>>,
 }
 
-impl<Rng: crate::Rng> TaskPool<Rng> {
+impl TaskPool {
```

Comment:
please return the `crate::Rng` trait, as we want to decide on the concrete random generator in `main`